### PR TITLE
Use SmallArray#

### PR DESCRIPTION
**Pull request by ethercrow**
As discussed here https://github.com/tibbe/unordered-containers/issues/126 it's now appropriate time for this PR.

### CODE REVIEW

**Comment by ethercrow** [comment]
Test suite fails to build with 8.2.

**Comment by ethercrow** [comment]
@nad Could you allow syb-0.7 in ChasingBottoms please?

**Comment by treeowl** [comment]
Ah, good point. I'll review as soon as I can.

**Comment by treeowl** [comment]
Do we really need to drop support for 7.8? How bad would it be to add a compat module with type synonyms and wrapper functions? The `SmallArray#` interface looks pretty much the same as the `Array#` one aside from names. I know we're not *obligated* to continue supporting 7.8, but I think we should only drop support for a GHC version when it becomes a substantial burden.

**Comment by ethercrow** [comment]
How about this way?

**Comment by ethercrow** [comment]
OK, this should be better.

**Comment by treeowl** [comment]
Sorry, I may not have been clear enough. I *do* think it makes good sense to drop support for super-old versions. I personally feel pretty comfortable dropping support for everything before 7.8 right now. To support 7.8, I think there should probably be a separate `SmallArray` module, so we only need to change the import line in `Array`.

**Comment by ethercrow** [comment]
OK dropped everything older than 7.8

File: Data/HashMap/Array.hs [hunk 1]
```diff
@@ -78,6 +81,24 @@ import qualified Prelude
 
 import Data.HashMap.Unsafe (runST)
 
+
+#if __GLASGOW_HASKELL__ >= 710
+type Array# a = SmallArray# a
+type MutableArray# a = SmallMutableArray# a
```
Comment (treeowl) [inline Data/HashMap/Array.hs]: You should probably use

```haskell
type Array# = SmallArray#
type MutableArray# = SmallMutableArray#
```

Not a big deal.

File: Data/HashMap/Base.hs [hunk 2]
```diff
@@ -1375,18 +1345,12 @@ fullNodeMask = complement (complement 0 `unsafeShiftL` maxChildren)
...
 
-#if __GLASGOW_HASKELL__ >= 708
 ------------------------------------------------------------------------
 -- IsList instance
 instance (Eq k, Hashable k) => Exts.IsList (HashMap k v) where
     type Item (HashMap k v) = (k, v)
     fromList = fromList
-    toList   = toList
-#endif
+    toList   = toList
```
Comment (treeowl) [inline Data/HashMap/Base.hs]: Please put in the missing newline.

**Comment by treeowl** [comment]
One more thing: please put the benchmark results in the commit message.

**Comment by treeowl** [comment]
Oh, blast! Did GHC get the type role wrong for `SmallArray#`???

**Comment by ethercrow** [comment]
And I'm getting the same error with GHC 8.2

**Comment by treeowl** [comment]
Oh, no, I was just wrong. You need the type synonyms to take arguments. Sorry. Dunno why...

**Comment by ethercrow** [comment]
Oh, I see

**Comment by treeowl** [comment]
I just opened https://ghc.haskell.org/trac/ghc/ticket/14101 to try to get this fixed. I can't imagine it's really supposed to work like that.

**Comment by ethercrow** [comment]
I added bench runs to the main commit. I didn't fix CPU frequency or isolated cores or anything like that, so the results are not too trustworthy.

**Comment by ethercrow** [comment]
Also funny and sad thing is, ghc-datasize seems to break on `SmallArray#`, so I didn't measure memory footprint.

**Comment by treeowl** [comment]
Ugh. Benchmarks that run in *microseconds*? The improvements are small, but so are the (maybe?) regressions. And it's pretty clear that card marking won't do a 16-element rarely-mutated array any good, so let's do this.

**Comment by treeowl** [comment]
Thanks, @ethercrow.

**Comment by ethercrow** [comment]
I forgot to delete now unused cbits directory.

**Comment by andrewthad** [comment]
I'm glad this has been merged, but I'd like to point out that users will no longer be able to add a hashmap to a compact region because the [SmallArray compaction code is not implemented yet](https://ghc.haskell.org/trac/ghc/ticket/13857).

**Pull request merged**